### PR TITLE
fix: Use corresponding sampler for segcore tracing

### PR DESCRIPTION
**Pull request by congqixia**
See also #35494

### CODE REVIEW

File: internal/core/src/common/Tracer.cpp [hunk 1]
```diff
@@ -84,8 +87,22 @@ initTelemetry(const TraceConfig& cfg) {
         resource::ResourceAttributes attributes = {
             {"service.name", TRACE_SERVICE_SEGCORE}, {"NodeID", cfg.nodeID}};
         auto resource = resource::Resource::Create(attributes);
-        auto sampler = std::make_unique<trace_sdk::ParentBasedSampler>(
-            std::make_shared<trace_sdk::AlwaysOnSampler>());
+
+        std::unique_ptr<trace_sdk::ParentBasedSampler> sampler;
+        if (cfg.sampleFraction == 0) {
```
Comment (PwzXxm) [inline internal/core/src/common/Tracer.cpp:92]: maybe compare with `std::numeric_limits<float>::epsilon`?

Comment (congqixia) [reply internal/core/src/common/Tracer.cpp]: fixed

**Comment by xiaofan-luan** [comment]
no i don't think this is the right fix.

Cpp sampler has to follow go. You can not use two sampler for cpp and go and all the important message will miss

**Comment by congqixia** [comment]
> no i don't think this is the right fix.
> 
> Cpp sampler has to follow go. You can not use two sampler for cpp and go and all the important message will miss

@xiaofan-luan this is what this PR do. The quickest way to sync go and cpp sampler is to use traceIDRatio sampling ,which is deterministic. Go & Cpp now share same policy to decide whether sample same trace id
plz check https://opentelemetry.io/docs/concepts/sampling/#head-sampling

The ultimate solution we want is to sample same trace after using tail sampling based on rt or error status, which shall be addressed in another PR.

**Comment by xiaofan-luan** [comment]
> > no i don't think this is the right fix.
> > Cpp sampler has to follow go. You can not use two sampler for cpp and go and all the important message will miss
> 
> @xiaofan-luan this is what this PR do. The quickest way to sync go and cpp sampler is to use traceIDRatio sampling ,which is deterministic. Go & Cpp now share same policy to decide whether sample same trace id plz check https://opentelemetry.io/docs/concepts/sampling/#head-sampling
> 
> The ultimate solution we want is to sample same trace after using tail sampling based on rt or error status, which shall be addressed in another PR.

I thought you need to bypass a signal in from go to cpp and indicate whether I need to trace this request.
using tracing ratio might be really hacky (what if not all the request goes into cpp?)

**Comment by congqixia** [comment]
> > > no i don't think this is the right fix.
> > > Cpp sampler has to follow go. You can not use two sampler for cpp and go and all the important message will miss
> > 
> > 
> > @xiaofan-luan this is what this PR do. The quickest way to sync go and cpp sampler is to use traceIDRatio sampling ,which is deterministic. Go & Cpp now share same policy to decide whether sample same trace id plz check https://opentelemetry.io/docs/concepts/sampling/#head-sampling
> > The ultimate solution we want is to sample same trace after using tail sampling based on rt or error status, which shall be addressed in another PR.
> 
> I thought you need to bypass a signal in from go to cpp and indicate whether I need to trace this request. using tracing ratio might be really hacky (what if not all the request goes into cpp?)

1. passing `sampling` flag is doable
2. traceID ratio is not hacky since the method is to "hashing" the trace id bytes, so if trace id and sample fraction is determined, the result is stable

Now we shall make a choice that which way we shall use to sync go & cpp trace sampling behavior.

Also another opening topic is how shall we implement tailing sampling with current design

**Comment by xiaofan-luan** [comment]
> > > > no i don't think this is the right fix.
> > > > Cpp sampler has to follow go. You can not use two sampler for cpp and go and all the important message will miss
> > > 
> > > 
> > > @xiaofan-luan this is what this PR do. The quickest way to sync go and cpp sampler is to use traceIDRatio sampling ,which is deterministic. Go & Cpp now share same policy to decide whether sample same trace id plz check https://opentelemetry.io/docs/concepts/sampling/#head-sampling
> > > The ultimate solution we want is to sample same trace after using tail sampling based on rt or error status, which shall be addressed in another PR.
> > 
> > 
> > I thought you need to bypass a signal in from go to cpp and indicate whether I need to trace this request. using tracing ratio might be really hacky (what if not all the request goes into cpp?)
> 
> 1. passing `sampling` flag is doable
> 2. traceID ratio is not hacky since the method is to "hashing" the trace id bytes (provided by opentelemetry official sdk), so if trace id and sample fraction is determined, the result is stable
> 
> Now we shall make a choice that which way we shall use to sync go & cpp trace sampling behavior.
> 
> Also another opening topic is how shall we implement tailing sampling with current design

1. I thought it make much sense to do tail sampling if the overhead is not super high (<10%). It make our life easier
2. since we already pass traceID through cgo, might be not that difficult  to pass sampling signal as well?

**Comment by congqixia** [comment]
@xiaofan-luan some updates:
1. tail sampling shall be done in another PR after some investigation
2. sample flag already passed via `traceFlags`, the `trace_sdk::ParentBasedSampler` will follow parent span sampling flag if parent ctx is valid

**Comment by PwzXxm** [comment]
Vote for tail sampling as well.
To my understanding, this PR only eliminates the runtime overhead where the tracer in CPP is always sampling things no matter what the sampling ratio is (if deterministic ofc). This design does not make sense in the first place.
We might fix this first and try to look into how to implement tail sampling.

**Pull request closed without merging**